add openziti's ziti and ziti-edge-tunnel programs

andrewzah opened the pull request:
This PR bumps the `ziti` version in #425154 (thanks @jamalhabash!) and adds `ziti-edge-tunnel` (thanks @rochecompaan!). These tools go hand in hand.

This closes #189696 and #425154.

## Things done:

- Built on platform:
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

jamalhabash committed:
maintainers: add jamalhabash

jamalhabash committed:
ziti: init at 2.0.2

see: https://github.com/openziti/ziti/releases/tag/v2.0.2

Co-authored-by: Andrew Zah <zah@andrewzah.com>

rochecompaan committed:
ziti-edge-tunnel: init at 1.18.2

See: https://github.com/openziti/ziti-tunnel-sdk-c

Co-authored-by: Andrew Zah <zah@andrewzah.com>

andrewzah commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review-gha`](https://github.com/Defelo/nixpkgs-review-gha)

Command: `nixpkgs-review pr 453502`
Commit: [`2db4b7a285c362643a0fdb44bfa4c954330f7a43`](https://github.com/NixOS/nixpkgs/commit/2db4b7a285c362643a0fdb44bfa4c954330f7a43) ([subsequent changes](https://github.com/NixOS/nixpkgs/compare/2db4b7a285c362643a0fdb44bfa4c954330f7a43..pull/453502/head))
Merge: [`7b38970aadc839d799322966dfe9f2a21860ff14`](https://github.com/NixOS/nixpkgs/commit/7b38970aadc839d799322966dfe9f2a21860ff14)

Logs: https://github.com/andrewzah/nixpkgs-review-gha/actions/runs/18629435831


---
### `x86_64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>ziti-edge-tunnel</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>ziti</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:x: 1 package failed to build:</summary>
  <ul>
    <li>ziti-edge-tunnel</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>ziti</li>
  </ul>
</details>

---
### `x86_64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>ziti</li>
  </ul>
</details>

---
### `aarch64-darwin` (sandbox = true)
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>ziti</li>
  </ul>
</details>

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+
+  nativeBuildInputs = [
+    cmake
+    git
```
```suggestion
```

Is Git really necessary here? The build seems to pass without it.

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+    mkdir -p patched-cmake
+    cp -r ${llhttp.dev}/lib/cmake/llhttp patched-cmake/
+    substituteInPlace patched-cmake/llhttp/llhttp-config.cmake \
+      --replace 'set(_IMPORT_PREFIX "${llhttp}")' 'set(_IMPORT_PREFIX "${llhttp.dev}")'
```
```suggestion
    cp -r ${lib.getDev llhttp}/lib/cmake/llhttp patched-cmake/
    substituteInPlace patched-cmake/llhttp/llhttp-config.cmake \
      --replace 'set(_IMPORT_PREFIX "${llhttp}")' 'set(_IMPORT_PREFIX "${lib.getDev llhttp}")'
```

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+  propagatedBuildInputs = [
+    systemd # For the resolvectl command at runtime
+  ];
+
```
```suggestion
  doInstallCheck = true;
  nativeInstallCheckInputs = [ versionCheckHook ];
  versionCheckProgramArg = "version";
```

Maybe we can add a `versionCheckHook` here to "test" the binary and check the version output.

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+{
+  pkgs,
```
```suggestion
```

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+    rev = "87350797774530b6ba9c00017f0f53dd57e6c38e";
+    hash = "sha256-Gz0/b9jcC1I0fmguSMkV0xiqKWq7vzUVT0Bd1F4iqkA=";
+  };
+  tlsuv_src = pkgs.fetchFromGitHub {
```
```suggestion
  tlsuv_src = fetchFromGitHub {
```

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+    "-DFETCHCONTENT_SOURCE_DIR_SUBCOMMAND=../deps/subcommand.c"
+    "-DFETCHCONTENT_SOURCE_DIR_TLSUV=../deps/tlsuv"
+    "-DDOXYGEN_OUTPUT_DIR=/tmp/doxygen"
+    "-DFETCHCONTENT_FULLY_DISCONNECTED=ON"
```
```suggestion
    (lib.cmakeBool "ENABLE_VCPKG" false)
    (lib.cmakeBool "DISABLE_SEMVER_VERIFICATION" true)
    (lib.cmakeBool "DISABLE_LIBSYSTEMD_FEATURE" true) # Disable direct integration to use resolvectl fallback
    (lib.cmakeFeature "ZITI_SDK_DIR" "${ziti_sdk_src}")
    (lib.cmakeFeature "ZITI_SDK_VERSION" "1.8.5")
    # Ensure a concrete version is embedded; upstream library stringifies ZITI_VERSION
    (lib.cmakeFeature "CMAKE_C_FLAGS" "-DZITI_VERSION=v${finalAttrs.version}")
    (lib.cmakeFeature "CMAKE_CXX_FLAGS" "-DZITI_VERSION=v${finalAttrs.version}")
    (lib.cmakeBool "FETCHCONTENT_FULLY_DISCONNECTED" true)
    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_LWIP" "${lwip_src}")
    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_LWIP-CONTRIB" "${lwip_contrib_src}")
    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_SUBCOMMAND" "${subcommand_c_src}")
    (lib.cmakeFeature "FETCHCONTENT_SOURCE_DIR_TLSUV" "${tlsuv_src}")
    (lib.cmakeFeature "DOXYGEN_OUTPUT_DIR" "/tmp/doxygen")
```

We can use the cmakeBool / cmakeFeature helpers here.

anthonyroussel commented on pkgs/by-name/zi/ziti-edge-tunnel/package.nix:
```diff
@@ -0,0 +1,133 @@
+    repo = "ziti-tunnel-sdk-c";
+    rev = "64900406faec1f02f59b73e3cdfd7e8299c40c76";
+    hash = "sha256-BOaXrttXDGAAzcGACpX9RQrlxTTdrU8IqdYJtZZY44o=";
+    fetchSubmodules = true;
```
```suggestion
```

Looks like there are no git submodules in the repository: https://github.com/openziti/ziti-tunnel-sdk-c/blob/v1.7.12/.gitmodules

andrewzah commented:
@anthonyroussel thank you for looking into this and giving feedback!

I believe I fixed the things you mentioned. The only thing is using `lwip`'s source breaks as it's read only, so I left it pointing to a relative path copied in `preConfigure`. Also, the version check command failed until I added ziti's version under `GIT_VERSION`.


```nix
cmakeFlags = [
  (cmakeFeature "ZITI_SDK_DIR" "${ziti_sdk_src}")
  (cmakeFeature "FETCHCONTENT_SOURCE_DIR_LWIP" "../../deps/lwip")
  (cmakeFeature "FETCHCONTENT_SOURCE_DIR_LWIP-CONTRIB" "${lwip_contrib_src}")
  (cmakeFeature "FETCHCONTENT_SOURCE_DIR_SUBCOMMAND" "${subcommand_c_src}")
  (cmakeFeature "FETCHCONTENT_SOURCE_DIR_TLSUV" "${tlsuv_src}")

  (cmakeFeature "CMAKE_C_FLAGS" "-DZITI_VERSION=v${finalAttrs.version}")
  (cmakeFeature "CMAKE_CXX_FLAGS" "-DZITI_VERSION=v${finalAttrs.version}")
  (cmakeFeature "CMAKE_C_FLAGS" "-DGIT_VERSION=${finalAttrs.version}")
  (cmakeFeature "CMAKE_CXX_FLAGS" "-DGIT_VERSION=${finalAttrs.version}")
];

```

```
CMake Error: Could not open file for write in copy operation /nix/store/z32mfgd710f5zz3p0h79g9w9x1yqfz16-source/src/include/lwip/init.h.tmp
CMake Error: : System Error: Permission denied
CMake Error at /nix/store/z32mfgd710f5zz3p0h79g9w9x1yqfz16-source/src/Filelists.cmake:256 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  lib/ziti-tunnel/CMakeLists.txt:39 (include)


CMake Error: Could not open file for write in copy operation /nix/store/z32mfgd710f5zz3p0h79g9w9x1yqfz16-source/doc/doxygen/lwip.Doxyfile.tmp
CMake Error: : System Error: Permission denied
CMake Error at /nix/store/z32mfgd710f5zz3p0h79g9w9x1yqfz16-source/src/Filelists.cmake:263 (configure_file):
  configure_file Problem configuring file
Call Stack (most recent call first):
  lib/ziti-tunnel/CMakeLists.txt:39 (include)
  ```
